prow/plugins/assign/assign.go

```diff
@@ -208,15 +240,122 @@ func newReviewHandler(e github.GenericCommentEvent, gc githubClient, log *logrus
 	addFailureResponse := func(mu github.MissingUsers) string {
 		return fmt.Sprintf("GitHub didn't allow me to request PR reviews from the following users: %s.\n\nNote that only [%s members](https://github.com/orgs/%s/people) and repo collaborators can review this PR, and authors cannot review their own PRs.", strings.Join(mu.Users, ", "), org, org)
 	}
+	addInvalidUsersResponse := func(invalidLogins []string) string {
+		return fmt.Sprintf("Cannot request PR reviews from the following users or teams: %s.\n\nI will request reviews from the rest of the specified users/teams if there are any",
+			strings.Join(invalidLogins, ", "))
+	}
 
 	return &handler{
-		addFailureResponse: addFailureResponse,
-		remove:             gc.UnrequestReview,
-		add:                gc.RequestReview,
-		event:              &e,
-		regexp:             CCRegexp,
-		gc:                 gc,
-		log:                log,
-		userType:           "reviewer(s)",
+		addFailureResponse:   addFailureResponse,
+		remove:               gc.UnrequestReview,
+		add:                  gc.RequestReview,
+		event:                &e,
+		regexp:               CCRegexp,
+		gc:                   gc,
+		log:                  log,
+		userType:             "reviewer(s)",
+		invalidLogins:        canReview(gc),
+		invalidUsersResponse: addInvalidUsersResponse,
+	}
+}
+
+func canAssign(gc githubClient) func(string, string, int, string, []string) ([]string, error) {
+	return func(org, repo string, number int, author string, users []string) ([]string, error) {
+		var invalidLogins []string
+		assignees, err := allowedAssignees(gc, org, repo, number, author)
+		if err != nil {
+			return []string{}, err
+		}
+		for _, user := range users {
+			if !assignees.Has(github.NormLogin(user)) {
+				invalidLogins = append(invalidLogins, user)
+			}
+		}
+		return invalidLogins, nil
+	}
+}
+
+func allowedAssignees(gc githubClient, org, repo string, number int, author string) (sets.String, error) {
+	assignees, err := gc.ListAssignees(org, repo)
+	if err != nil {
+		return sets.String{}, err
+	}
+	comments, err := gc.ListIssueComments(org, repo, number)
+	if err != nil {
+		return sets.String{}, err
+	}
+
+	allAssignees := sets.NewString(logins(assignees)...)
+	allAssignees.Insert(github.NormLogin(author))
+	allAssignees.Insert(commenters(comments)...)
```

Comment:
Huh, this is strange - I would have thought that the [API under `ListAssignees`](https://developer.github.com/v3/issues/assignees/#list-assignees) already includes the author and commenters in the set. It doesn't? Maybe worth a comment...

Reply:
Thank you for the review. Based on a discussion with @stevekuznetsov I updated the code to handle 404 from GitHub API, rather than using tokens to build a list of valid assignees.